adapters/adnuntius/adnuntius.go

```diff
@@ -0,0 +1,300 @@
+		}
+	}
+}
+type adnMetaData struct {
+	Usi string `json:"usi,omitempty"`
+}
+type adnRequest struct {
+	AdUnits  []adnAdunit `json:"adUnits"`
+	MetaData adnMetaData `json:"metaData,omitempty"`
+	Context  string      `json:"context,omitempty"`
+}
+
+// Builder builds a new instance of the Adnuntius adapter for the given bidder with the given config.
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter) (adapters.Bidder, error) {
+	bidder := &adapter{
+		tzo:      realTzo{},
+		endpoint: config.Endpoint,
+	}
+
+	return bidder, nil
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var extRequests []*adapters.RequestData
+	var errs []error
+
+	extRequests, err := a.generateRequests(*request)
+	if err != nil {
+		errs = append(errs, err...)
+	}
+	return extRequests, errs
+}
+
+func setHeaders() http.Header {
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	return headers
+}
+
+func makeEndpointUrl(ortbRequest openrtb2.BidRequest, a *adapter) (string, []error) {
+	uri, err := url.Parse(a.endpoint)
+	if err != nil {
+		return "", []error{fmt.Errorf("failed to parse Adnuntius endpoint: %v", err)}
+	}
+
+	gdpr, consent, err := getGDPR(&ortbRequest)
+	if err != nil {
+		return "", []error{fmt.Errorf("failed to parse Adnuntius endpoint: %v", err)}
+	}
+
+	_, offset := a.tzo.Now().UTC().Local().Zone()
+	tzo := -offset / 3600 * 60
+
+	q := uri.Query()
+	if gdpr != "" && consent != "" {
+		q.Set("gdpr", gdpr)
+		q.Set("consentString", consent)
+	}
+	q.Set("tzo", fmt.Sprint(tzo))
+	q.Set("format", "json")
+
+	url := a.endpoint + "?" + q.Encode()
+	return url, nil
+}
+
+/*
+	Generate the requests to Adnuntius to reduce the amount of requests going out.
+*/
+func (a *adapter) generateRequests(ortbRequest openrtb2.BidRequest) ([]*adapters.RequestData, []error) {
+	var requestData []*adapters.RequestData
+	var errors []error
+	networkAdunitMap := make(map[string][]adnAdunit)
+	headers := setHeaders()
+
+	endpoint, err := makeEndpointUrl(ortbRequest, a)
+	if err != nil {
+		errors = append(errors, fmt.Errorf("failed to parse URL: %v", err))
+	}
+
+	for _, imp := range ortbRequest.Imp {
+		if imp.Banner == nil {
+			errors = append(errors, &errortypes.BadInput{
+				Message: fmt.Sprintf("ignoring imp id=%s, Adnuntius supports only Banner", imp.ID),
+			})
+		}
+		var bidderExt adapters.ExtImpBidder
+		if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+			errors = append(errors, &errortypes.BadInput{
+				Message: fmt.Sprintf("Error unmarshalling ExtImpBidder: %s", err.Error()),
+			})
+		}
+
+		var adnuntiusExt openrtb_ext.ImpExtAdnunitus
+		if err := json.Unmarshal(bidderExt.Bidder, &adnuntiusExt); err != nil {
+			errors = append(errors, &errortypes.BadInput{
+				Message: fmt.Sprintf("Error unmarshalling ExtImpBmtm: %s", err.Error()),
+			})
+		}
```

Comment:
In case of error occurs, lines 128, 134 and 141 - should you just return an error and abort the execution?

Reply:
Hopefully I've done it as you meant it in my coming commit.

Reply:
Thank you for the update.
Please check it one more time. You probably don't want to collect all errors. Instead return after first error occurrence. If so please modify return `[]error` to just `error` and return it immediately.